memtrack-v5

precision-software opened the pull request.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,36 @@ include_dir 'conf.d'
+      <listitem>
+       <para>
+        Specifies a limit to the amount of memory (MB) that may be allocated to
+        backends in total (i.e. this is not a per user or per backend limit).
```
How about "by PostgreSQL in total" or similar, rather than talking about "backends" here?

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,36 @@ include_dir 'conf.d'
+        Specifies a limit to the amount of memory (MB) that may be allocated to
+        backends in total (i.e. this is not a per user or per backend limit).
+        If unset, or set to 0 it is disabled.  At databse startup
+        max_total_memory is reduced by shared_memory_size_mb
```
This would be better written in the positive form- that is "included in the memory accounting is shared_buffers, other shared memory which is needed for PostgreSQL, memory allocated by individual PostgreSQL backends, and memory allocated by PostgreSQL auxiliary processes.  Note that processes spawned by PostgreSQL during operation (eg: archive_command), extensions or libraries which don't use PostgreSQL's memory accounting system (palloc), startup processes which run before PostgreSQL starts and client utilities, will not be accounted for and therefore this value should probably be configured a bit less than the actual maximum memory available in the system.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,36 @@ include_dir 'conf.d'
+        setting this value. A backend request that would exhaust the limit will
+        be denied with an out of memory error causing that backend's current
+        query/transaction to fail. Further requests will not be allocated until
+        dropping below the limit.  This limit does not affect auxiliary backend
```
This is making me think we might need to have some kind of 'reserved' memory for auxiliary backends...

precision-software replied:
Not hard to implement or to add later. We would pick different upper limits depending on the type of process.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,36 @@ include_dir 'conf.d'
+        query/transaction to fail. Further requests will not be allocated until
+        dropping below the limit.  This limit does not affect auxiliary backend
+        processes
+        <xref linkend="glossary-auxiliary-proc"/> or the postmaster process.
```
I wonder how much memory the postmaster really allocates once it's up and running?

sfrost commented on src/backend/storage/ipc/dsm_impl.c:
```diff
@@ -155,12 +157,12 @@ int			min_dynamic_shared_memory;
-			void **impl_private, void **mapped_address, Size *mapped_size,
-			int elevel)
+static bool
+original_dsm_impl_op(dsm_op op, dsm_handle handle, Size request_size,
```
"original" really isn't a good name for this- we don't want to end up with three levels of "original" or "prior" or such- the code should really look like it was all written at once.  We tend to call this kinds of functions "dsm_impl_op_guts" or "dsm_impl_op_worker" or similar.

precision-software replied:
changing to "worker", as in dsm_impl_op_worker.

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,282 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+  </indexterm>
+
+  <para>
+   The <structname>pg_stat_global_memory_allocation</structname> view will
```
Presuming we make the changes to remove the extra columns from this view that we discussed previously, the documentation for the view also needs to be updated.

precision-software replied:
Definitely.  It's not an onerous change, but I'm saving it for last.  

That "hack" with saving mapped_size was economical, but not a clean solution. It also missed the case where we destroy a segment we're not attached to, say in "unpin". The update saves requested size in dsm_control so it fixes the unpin case as well.

precision-software replied:
Oops - I was saving adding the "INIT" column for last. I'm working on the other call as we speak, so thanks for the reminder.

precision-software replied:
This was an oops - I was responding to a different comment.  I hadn't made those changes yet. I have removed those columns from the docs and still need to review the docs for proper wordsmithing.

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,185 @@
+#include "storage/proc.h"
+
+/* Forward references */
+static inline bool atomic_add_with_limit(pg_atomic_uint64 *sum, uint64 add, uint64 limit);
```
Thought we agreed to move this into the atomics header file?

precision-software replied:
I'm all in favor of putting it in atomics.h, if we can introduce just that one routine. 

If we need to introduce a complete set of variations like PG has for atomic add, we would need 8 new routines
 - pg_atomic_fetch_add_limit_u64, pg_atomic_add_fetch_limit_u64, 
 - pg_atomic_fetch_sub_limit_u64, pg_atomic_sub_fetch_limit_u64
 -  pg_atomic_fetch_add_limit_u32, pg_atomic_add_fetch_limit_u32, 
 - pg_atomic_fetch_sub_limit_u32, pg_atomic_sub_fetch_limit_u32
 along with a new set of unit tests.

In the latter case, I'd rather hold off moving it to atomics.h.  Your thoughts?

sfrost replied:
Let's start with just introducing the one routine and see what folks say.

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,185 @@
+	if (ProcGlobal == NULL || MyProcPid == 0)
+		return update_local_allocation(size, type);
+
+	/* Quick check to verify totals are not negative. Should never happen. */
```
Should say "Assert to ...", doesn't matter if it's quick or not, it won't be in a production build.

precision-software replied:
I was thinking of dropping the preface completely and saying "Verify totals are not negative. Should never happen.".

sfrost replied:
Sure, that's fine too.

sfrost commented on src/backend/postmaster/postmaster.c:
```diff
@@ -6326,6 +6329,9 @@ restore_backend_variables(BackendParameters *param, Port *port)
 	strlcpy(pkglib_path, param->pkglib_path, MAXPGPATH);
 
+	max_total_memory_mb = param->max_total_memory_mb;
+	max_total_bkend_bytes = (uint64)max_total_memory_mb * 1024 * 1024;
```
Surely this is now max_total_memory_bytes..?  Surprised this even compiled..

precision-software replied:
done.

sfrost commented on src/backend/storage/ipc/ipci.c:
```diff
@@ -359,4 +359,29 @@ InitializeShmemGUCs(void)
+					 errhint("Disable or increase the configuration parameter \"max_total_memory\"."));
+
+		 /* Warning if less than 100MB available for non-shared memory */
+		 if (max_total_memory_mb - size_mb < 100)
```
Thought we agreed to change this to being reflective of the max_connections setting?

precision-software replied:
Working my way down the list.  I'm not completely happy with the changed version, so it still needs a bit of attention.

sfrost reviewed (changes requested):
Looks like there's still a lot of changes from our discussion yesterday that are needed?  Tried to note down the ones that I remember from what you were adding to your todo list.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,37 @@ include_dir 'conf.d'
+      <listitem>
+       <para>
+        Specifies a limit to the total amount of memory (MB) that may be allocated to
+        the postgres server. (i.e. this is not a per user or per backend limit).
```
Looking at other documentation, eg: max_connections, seems we typically refer to this as "database server".  Please update to that terminology.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,37 @@ include_dir 'conf.d'
+      </term>
+      <listitem>
+       <para>
+        Specifies a limit to the total amount of memory (MB) that may be allocated to
```
I think "allocated by" is better wording here

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,37 @@ include_dir 'conf.d'
+        Specifies a limit to the total amount of memory (MB) that may be allocated to
+        the postgres server. (i.e. this is not a per user or per backend limit).
+        If unset, or set to 0 it is disabled.  At database startup,
+        max_total_memory is reduced by shared_memory_size
```
I don't think "reduced by" is the right language here.

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,206 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+      allocated by the server, along with a fixed amount of memory
+      for each server process. Due to the dynamic nature of memory allocation,
+      this total is not exact, but it should be sufficient for monitoring purposes.
+      <xref linkend="guc-max-total-backend-memory"/>.
```
This isn't the right GUC anymore and why is it being referenced here anyway?  This is really independent, after all...

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,206 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+      </para>
+     <para>
+      Reports the amount of shared memory (in bytes) currently used by the server.
+      <xref linkend="guc-max-total-backend-memory"/>.
```
Same as above comment.

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,206 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+      cancellation of the process's active query/transaction.
+      If memory is not being limited, this column returns -1.
+      TODO: should it be -1 or NULL?
+      <xref linkend="guc-max-total-backend-memory"/>.
```
This one might make sense though I'm not completely sure, the GUC name should still be updated...

sfrost commented on src/backend/catalog/system_views.sql:
```diff
@@ -1344,4 +1344,10 @@ CREATE VIEW pg_stat_subscription_stats AS
 
 CREATE VIEW pg_wait_events AS
-    SELECT * FROM pg_get_wait_events();
+SELECT * FROM pg_get_wait_events();
```
Please fix the indentation here so this doesn't look like a change, since we aren't intending to change this...

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,33 @@ include_dir 'conf.d'
+        but the postmaster is able to exceed the limit.
+        Total memory used (<varname>total_memory_used</varname>) is displayed in the
+        <link linkend="monitoring-pg-stat-memory-allocation-view">
+        <structname>pg_stat_memory_allocation</structname></link view.
```
Missing the closing > in /link

sfrost commented on src/backend/storage/ipc/dsm.c:
```diff
@@ -14,6 +14,12 @@
  * still exist, at the next postmaster startup.
  *
+ * When invoking the low level functions, the caller must correctly
+ * set "request_size" when creating and destroying segments.
```
Really not sure if this is going to end up flying with the community.  Also- I'm not 100% sure that we're always going to know the size before destroying, if it isn't the same process creating as destroying?  If we do know, then we must be keeping track of that somewhere and, if we're doing that, why can't we just pull that info?

precision-software replied:
Currently, we don't seem to be keeping the info around. We do keep "mapped size" for each mapped segment, but we have a problem when deleting a segment we haven't mapped (eg. unpin). Part of this change is to stash creation size in the dsm table where we can easily access it.

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+
+	/* Verify totals are not negative. Should never happen. */
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->total_memory_bytes) >= 0);
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->shared_memory_bytes) >= 0);
```
This assert is failing when we're going through a crash-restart cycle due to a backend getting hit with a kill -9.  You can see this by doing a make check-world and/or just going into src/test/recovery and running 'make check' with TAP tests enabled.  Here's what the end of the PG server crash/restart log looks like, in case it's helpful:


> 2023-09-13 13:58:27.476 EDT [215504] LOG:  server process (PID 215533) was terminated by signal 9: Killed
> 2023-09-13 13:58:27.476 EDT [215504] DETAIL:  Failed process was running: INSERT INTO alive VALUES(in-progress-before-sigkill) RETURNING status;
> 2023-09-13 13:58:27.476 EDT [215504] LOG:  terminating any other active server processes
> 2023-09-13 13:58:27.478 EDT [215504] LOG:  all server processes terminated; reinitializing
> TRAP: failed Assert("(int64) pg_atomic_read_u64(&ProcGlobal->shared_memory_bytes) >= 0"), File: "src/backend/utils/activity/memtrack.c", Line: 162, PID: 215504
> bin/postgres(ExceptionalCondition+0x6e)[0x55ebf9f30726]
> bin/postgres(exit_tracked_memory+0x0)[0x55ebf9deee75]
> bin/postgres(dsm_impl_op+0x19f)[0x55ebf9da6924]
> bin/postgres(+0x4ddece)[0x55ebf9da3ece]
> bin/postgres(shmem_exit+0xf2)[0x55ebf9da6a18]
> bin/postgres(+0x467670)[0x55ebf9d2d670]
> bin/postgres(+0x468854)[0x55ebf9d2e854]
> bin/postgres(+0x469051)[0x55ebf9d2f051]
> bin/postgres(BackgroundWorkerInitializeConnection+0x0)[0x55ebf9d305f9]
> bin/postgres(main+0x228)[0x55ebf9c47216]
> /lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f2a6fc29d90]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f2a6fc29e40]
> bin/postgres(_start+0x25)[0x55ebf99a0445]

sfrost replied:
What I now suspect is happening is that we are now in a state where we *think* we can happily allocate additional memory (up to 1MB, all of which might be DSM) without reporting- but what we have actually stored into shared_memory_bytes is not "all current allocations, plus 1MB of space to let us allocate more" but rather "all current DSM allocations", which means that as soon as we allocate additional DSM space, some other process could come along and de-allocate that space (and maybe all other DSM allocations) and that ends up pushing this value negative.

Always updating the shared memory value for every allocation is one way to address this, as proposed, though the reasoning should be explained clearly if that's the direction we decide to go in.

Alternatively, we could bump the initial value for shared memory by 1MB ... but I would want to wait to do that until the first actual DSM allocation that this process does rather than do it always (in the way we do for the general total memory allocation).

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,200 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+      Reports the total amount of memory (in bytes) currently used by the server.
+      This total includes shared memory segments, actual memory
+      allocated by the server, along with a fixed amount of memory
+      assumed for each server process. Due to the dynamic nature of memory allocation,
```
This in't just a fixed amount of memory but is actually how much memory we are currently tracking for each backend, right..?  Plus maybe 1MB that we kind of assume is what's needed per-backend that just gets copied as part of being modified post-fork, but I don't think I'd even mention that here- it's covered in the "this isn't exact" comment, imv.  Certainly part of the idea here is that folks can monitor this value vs. the max-total-memory value to see if they're getting close to running into that limit and having queries start failing...

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,200 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+       <structfield>datid</structfield> <type>oid</type>
+      </para>
+      <para>
+       OID of the database this backend is connected to
```
Why do we have this...?  Nothing in what we're talking about here is really per-database, so the OID of the database we're connected to doesn't seem useful...

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+int64		allocation_upper_bound = 0;
+
+/* ---------
+ * init_backend_memory() -
```
Comment needs to be updated.

sfrost replied:
... still needs to be updated, it seems :)

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+	 */
+	dsm_delta = my_memory.allocated_bytes_by_type[PG_ALLOC_DSM] -
+		reported_memory.allocated_bytes_by_type[PG_ALLOC_DSM];
+	pg_atomic_fetch_add_u64(&ProcGlobal->shared_memory_bytes, dsm_delta);
```
Surely we should only be doing this if dsm_delta != 0?

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+	my_memory = INIT_BACKEND_MEMORY;
+	reported_memory = NO_BACKEND_MEMORY;
+
+	/* Force the next allocation to do global bounds checking. */
```
Have to go through update_global_allocations first time post-shared-memory init

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+	if (ProcGlobal == NULL || MyProcPid == 0)
+		return update_local_allocation(size, type);
+
+	/* Verify totals are not negative. Should never happen. */
```
First time reaching here, my_memory will have 1MB+pre-shmmem-init-allocations, reported will be 0, limit will 0.

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+	 * Update the private memory counters. This must happen after the limit is
+	 * checked.
+	 */
+	update_local_allocation(size, type);
```
Here we update the local allocations to include 'size'

sfrost commented on src/backend/utils/activity/memtrack.c:
```diff
@@ -0,0 +1,165 @@
+	 * Update the global dsm memory counter to reflect changes since our last
+	 * report
+	 */
+	dsm_delta = my_memory.allocated_bytes_by_type[PG_ALLOC_DSM] -
```
Unlike with the total, we never initialize allocated_bytes_by_type to a non-zero value, so here it would be whatever 'size' is (since we did include that during updat_local_allocation).

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,200 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+       <structfield>shared_memory_used</structfield> <type>bigint</type>
+      </para>
+     <para>
+      Reports the amount of shared memory (in bytes) currently used by the server.
```
Need to adjust- we should be clear that this is reporting DSM (dynamic shared memory) and does not include the global shared memory (shared buffers, et al).

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,34 @@ include_dir 'conf.d'
+        Specifies a limit to the total amount of memory (MB) a database server
+        can allocate. (i.e. this is not a per user or per backend limit).
+        If unset or set to 0, it is disabled. max_total_memory must be larger than
+        shared_memory_size and should include an additional 2MB per connection.
```
I don't like how this is saying "should include" ... maybe "at least include" or similar, or we could just drop that part of the sentence entirely.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,34 @@ include_dir 'conf.d'
+        <para>
+        Any memory request that would exhaust the limit will
+        be denied with an out of memory error, causing that process's current
+        query/transaction to fail. Further requests will not be allocated until
```
s/allocated/allowed/ would be better wording in documentation, I think.

sfrost commented on doc/src/sgml/config.sgml:
```diff
@@ -2145,6 +2145,34 @@ include_dir 'conf.d'
+        but the postmaster is able to exceed the limit.
+        Total memory used (<varname>total_memory_used</varname>) is displayed in the
+        <link linkend="monitoring-pg-stat-memory-allocation-view">
+        <structname>pg_stat_memory_allocation</structname></link> view.
```
This should be saying pg-stat-global-memory-allocation, since the total isn't in the pg-stat-memory-allocation view.

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,190 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+      </para>
+     <para>
+      Total memory currently allocated to this backend.
+      It is the balance of bytes allocated and freed.
```
This sentence seems unnecessary?

sfrost commented on doc/src/sgml/monitoring.sgml:
```diff
@@ -4522,6 +4522,190 @@ description | Waiting for a newly initialized WAL file to reach durable storage
+       <structfield>dsm_allocated_bytes</structfield> <type>bigint</type>
+      </para>
+      <para>
+       Subtotal representing dynamic shared memory (dsm) created by this backend.
```
Seems like we should say 'allocated' here too, to match the other parameters, not 'created'.

sfrost commented on src/backend/storage/ipc/ipci.c:
```diff
@@ -359,4 +362,45 @@ InitializeShmemGUCs(void)
+	 * of shared memory. This is the first place where we know the size of
+	 * shmem, so it is also the first place we can validate the memory limit.
+	 */
+	if (max_total_memory_mb > 0)
```
We had discussed moving this check..?  Just wondering what happened to that change or if there was a reason you decided we should keep this here.

sfrost commented on src/backend/utils/activity/backend_status.c:
```diff
@@ -49,7 +50,6 @@ int			pgstat_track_activity_query_size = 1024;
 /* exposed so that backend_progress.c can access it */
 PgBackendStatus *MyBEEntry = NULL;
 
-
```
I know it's probably pgindent's fault, but please remove the whitespace-only changes.

precision-software commented:
Stephen,

There seems to have been a merge problem, or more likely I uploaded the wrong patch file last night. I’ll get back after breakfast.


  *   J

From: Stephen Frost ***@***.***>
Reply-To: precision-software/postgres ***@***.***>
Date: Friday, September 8, 2023 at 7:18 AM
To: precision-software/postgres ***@***.***>
Cc: John Morris ***@***.***>, Author ***@***.***>
Subject: Re: [precision-software/postgres] memtrack-v5 (PR #12)


@sfrost requested changes on this pull request.

Looks like there's still a lot of changes from our discussion yesterday that are needed? Tried to note down the ones that I remember from what you were adding to your todo list.

________________________________

In doc/src/sgml/config.sgml<https://github.com/precision-software/postgres/pull/12#discussion_r1319880567>:

> @@ -2145,6 +2145,36 @@ include_dir 'conf.d'

       </listitem>

      </varlistentry>



+     <varlistentry id="guc-max-total-backend-memory" xreflabel="max_total_memory">

+      <term><varname>max_total_memory</varname> (<type>integer</type>)

+      <indexterm>

+       <primary><varname>max_total_memory</varname> configuration parameter</primary>

+      </indexterm>

+      </term>

+      <listitem>

+       <para>

+        Specifies a limit to the amount of memory (MB) that may be allocated to

+        backends in total (i.e. this is not a per user or per backend limit).

How about "by PostgreSQL in total" or similar, rather than talking about "backends" here?

________________________________

In doc/src/sgml/config.sgml<https://github.com/precision-software/postgres/pull/12#discussion_r1319890535>:

> @@ -2145,6 +2145,36 @@ include_dir 'conf.d'

       </listitem>

      </varlistentry>



+     <varlistentry id="guc-max-total-backend-memory" xreflabel="max_total_memory">

+      <term><varname>max_total_memory</varname> (<type>integer</type>)

+      <indexterm>

+       <primary><varname>max_total_memory</varname> configuration parameter</primary>

+      </indexterm>

+      </term>

+      <listitem>

+       <para>

+        Specifies a limit to the amount of memory (MB) that may be allocated to

+        backends in total (i.e. this is not a per user or per backend limit).

+        If unset, or set to 0 it is disabled.  At databse startup

+        max_total_memory is reduced by shared_memory_size_mb

This would be better written in the positive form- that is "included in the memory accounting is shared_buffers, other shared memory which is needed for PostgreSQL, memory allocated by individual PostgreSQL backends, and memory allocated by PostgreSQL auxiliary processes. Note that processes spawned by PostgreSQL during operation (eg: archive_command), extensions or libraries which don't use PostgreSQL's memory accounting system (palloc), startup processes which run before PostgreSQL starts and client utilities, will not be accounted for and therefore this value should probably be configured a bit less than the actual maximum memory available in the system.

________________________________

In doc/src/sgml/config.sgml<https://github.com/precision-software/postgres/pull/12#discussion_r1319893028>:

> +       <primary><varname>max_total_memory</varname> configuration parameter</primary>

+      </indexterm>

+      </term>

+      <listitem>

+       <para>

+        Specifies a limit to the amount of memory (MB) that may be allocated to

+        backends in total (i.e. this is not a per user or per backend limit).

+        If unset, or set to 0 it is disabled.  At databse startup

+        max_total_memory is reduced by shared_memory_size_mb

+        (includes shared buffers and other memory required for initialization).

+        Each backend process is intialized with a 1MB local allowance which

+        also reduces total_memory_bytes_available. Keep this in mind when

+        setting this value. A backend request that would exhaust the limit will

+        be denied with an out of memory error causing that backend's current

+        query/transaction to fail. Further requests will not be allocated until

+        dropping below the limit.  This limit does not affect auxiliary backend

This is making me think we might need to have some kind of 'reserved' memory for auxiliary backends...

________________________________

In doc/src/sgml/config.sgml<https://github.com/precision-software/postgres/pull/12#discussion_r1319894065>:

> +      </term>

+      <listitem>

+       <para>

+        Specifies a limit to the amount of memory (MB) that may be allocated to

+        backends in total (i.e. this is not a per user or per backend limit).

+        If unset, or set to 0 it is disabled.  At databse startup

+        max_total_memory is reduced by shared_memory_size_mb

+        (includes shared buffers and other memory required for initialization).

+        Each backend process is intialized with a 1MB local allowance which

+        also reduces total_memory_bytes_available. Keep this in mind when

+        setting this value. A backend request that would exhaust the limit will

+        be denied with an out of memory error causing that backend's current

+        query/transaction to fail. Further requests will not be allocated until

+        dropping below the limit.  This limit does not affect auxiliary backend

+        processes

+        <xref linkend="glossary-auxiliary-proc"/> or the postmaster process.

I wonder how much memory the postmaster really allocates once it's up and running?

________________________________

In src/backend/storage/ipc/dsm_impl.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319904183>:

> @@ -155,12 +157,12 @@ int                   min_dynamic_shared_memory;

  * silently return false.

  *-----

  */

-bool

-dsm_impl_op(dsm_op op, dsm_handle handle, Size request_size,

-                      void **impl_private, void **mapped_address, Size *mapped_size,

-                      int elevel)

+static bool

+original_dsm_impl_op(dsm_op op, dsm_handle handle, Size request_size,

"original" really isn't a good name for this- we don't want to end up with three levels of "original" or "prior" or such- the code should really look like it was all written at once. We tend to call this kinds of functions "dsm_impl_op_guts" or "dsm_impl_op_worker" or similar.

________________________________

In src/backend/catalog/system_views.sql<https://github.com/precision-software/postgres/pull/12#discussion_r1319905682>:

> @@ -1343,5 +1343,41 @@ CREATE VIEW pg_stat_subscription_stats AS

     FROM pg_subscription as s,

          pg_stat_get_subscription_stats(s.oid) as ss;



+CREATE VIEW pg_stat_memory_allocation AS

+    SELECT

+        S.datid AS datid,

+        S.pid,

+        S.allocated_bytes,

+        S.aset_allocated_bytes,

+        S.dsm_allocated_bytes,

+        S.generation_allocated_bytes,

+        S.slab_allocated_bytes

+    FROM pg_stat_get_memory_allocation(NULL) AS S

+        LEFT JOIN pg_database AS D ON (S.datid = D.oid);

+

 CREATE VIEW pg_wait_events AS

Thought we agreed to move this to below pg_stat_global_memory_allocation ?

________________________________

In src/backend/catalog/system_views.sql<https://github.com/precision-software/postgres/pull/12#discussion_r1319906549>:

>  CREATE VIEW pg_wait_events AS

-    SELECT * FROM pg_get_wait_events();

+    SELECT * FRom pg_get_wait_events();

+

+CREATE VIEW pg_stat_global_memory_allocation AS

Thought we agreed to remove the summations and various other things from this?

Though I do wonder if we want to have another view that brings everything together? Not sure. Maybe we keep it simple for now.

________________________________

In doc/src/sgml/monitoring.sgml<https://github.com/precision-software/postgres/pull/12#discussion_r1319907213>:

> +

+    </tbody>

+   </tgroup>

+  </table>

+

+ </sect2>

+

+ <sect2 id="monitoring-pg-stat-global-memory-allocation-view">

+  <title><structname>pg_stat_global_memory_allocation</structname></title>

+

+  <indexterm>

+   <primary>pg_stat_global-memory_allocation</primary>

+  </indexterm>

+

+  <para>

+   The <structname>pg_stat_global_memory_allocation</structname> view will

Presuming we make the changes to remove the extra columns from this view that we discussed previously, the documentation for the view also needs to be updated.

________________________________

In src/backend/postmaster/postmaster.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319907617>:

> @@ -539,6 +539,7 @@ typedef struct

 #endif

        char           my_exec_path[MAXPGPATH];

        char           pkglib_path[MAXPGPATH];

+       int                    max_total_memory_mb;

Still need to get that pgindent run working.

________________________________

In src/backend/utils/activity/memtrack.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319928594>:

> + * IDENTIFICATION

+ *       src/backend/utils/activity/memtrack.c

+ *

+ *-------------------------------------------------------------------------

+ */

+

+#include <unistd.h>

+

+#include "postgres.h"

+#include "miscadmin.h"

+#include "utils/backend_status.h"

+#include "utils/memtrack.h"

+#include "storage/proc.h"

+

+/* Forward references */

+static inline bool atomic_add_with_limit(pg_atomic_uint64 *sum, uint64 add, uint64 limit);

Thought we agreed to move this into the atomics header file?

________________________________

In src/backend/utils/activity/memtrack.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319929930>:

> + * global values and pgstat statistics.

+ *

+ * Note the use of pg_atomic_fetch_add_u64() to update counters

+ * with a signed value assumes we are on a 2's complement machine.

+ * Otherwise, we would need to introduce atomic s64 operators.

+ */

+bool update_global_allocation(int64 size, pg_allocator_type type)

+{

+       int64 delta;

+       int64 dsm_delta;

+

+       /* If we are still initializing, only update the private counters */

+       if (ProcGlobal == NULL || MyProcPid == 0)

+              return update_local_allocation(size, type);

+

+       /* Quick check to verify totals are not negative. Should never happen. */

Should say "Assert to ...", doesn't matter if it's quick or not, it won't be in a production build.

________________________________

In src/backend/postmaster/postmaster.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319930913>:

> @@ -6326,6 +6329,9 @@ restore_backend_variables(BackendParameters *param, Port *port)



        strlcpy(pkglib_path, param->pkglib_path, MAXPGPATH);



+       max_total_memory_mb = param->max_total_memory_mb;

+       max_total_bkend_bytes = (uint64)max_total_memory_mb * 1024 * 1024;

Surely this is now max_total_memory_bytes..? Surprised this even compiled..

________________________________

In src/backend/storage/ipc/ipci.c<https://github.com/precision-software/postgres/pull/12#discussion_r1319932624>:

> +     /*

+        * Validating the backend memory limit GUC depends on knowing the size of

+        * shared memory. This is the first place where we know the size of shmem,

+        * so it is also the first place we can validate the backend memory limit.

+        */

+     if (max_total_memory_mb > 0)

+        {

+               /* Error if backend memory limit is less than shared memory size */

+               if (max_total_memory_mb <= size_mb)

+                       ereport(ERROR,

+                                      errmsg("configured max_total_memory %dMB is <= shared_memory_size %ldMB",

+                                                    max_total_memory_mb, size_mb),

+                                      errhint("Disable or increase the configuration parameter \"max_total_memory\"."));

+

+               /* Warning if less than 100MB available for non-shared memory */

+               if (max_total_memory_mb - size_mb < 100)

Thought we agreed to change this to being reflective of the max_connections setting?

—
Reply to this email directly, view it on GitHub<https://github.com/precision-software/postgres/pull/12#pullrequestreview-1617508365>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABLMUH6SFQYADMZCO6VET4TXZMSNHANCNFSM6AAAAAA4MYBOF4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

precision-software commented:
*   There seems to have been a merge problem, or more likely I uploaded the wrong patch file last night. I’ll get back after breakfast.

I was reading the oldest email in the thread, not the latest.  Now that I’ve had my morning latte’ I’ll get onto the changes.


  *   John

precision-software committed:
Updated patch to manage Postgres memory allocations.
  - Tracks memory allocations for DSM and Postgres allocators
  - Reserves total memory for server and limits how much memory a server can allocate
  - Provides views to show current memory utilization for individual processes and the server as a whole
